src/builtin/filter.rs

```diff
@@ -1,58 +1,72 @@
+use chrono::{DateTime, NaiveDateTime, Utc};
+use crate::engine::context::Context;
 use crate::error::Result;
 use serde_json::Value;
+use slug;
+use std::collections::HashMap;
 
-pub type FilterFn = fn(&Value) -> Result<Value>;
+pub type FilterFn = fn(&Value, &HashMap<String, Value>, &Context) -> Result<Value>;
 
-pub(crate) fn lower(value: &Value) -> Result<Value> {
+pub(crate) fn lower(value: &Value, _args: &HashMap<String, Value>, _context: &Context) -> Result<Value> {
     let s = value.as_str().ok_or_else(|| "`lower` filter accepts string only")?;
     Ok(Value::String(s.to_lowercase()))
 }
 
-pub(crate) fn upper(value: &Value) -> Result<Value> {
+pub(crate) fn upper(value: &Value, _args: &HashMap<String, Value>, _context: &Context) -> Result<Value> {
     let s = value.as_str().ok_or_else(|| "`upper` filter accepts string only")?;
     Ok(Value::String(s.to_uppercase()))
 }
 
-#[cfg(test)]
-mod tests {
-    use super::{lower, upper};
-    use serde_json::json;
+fn format_timestamp(filter: &str, value: &Value, args: &HashMap<String, Value>, default: &str) -> Result<Value> {
```

Comment:
as the `value, args` and often `context` seem to travel together in the signatures of these functions - would it make sense to introduce a type or a type alias for them ?

Reply:
Maybe, in my experience, whenever I do something like this, I need to break it after couple of days, because there's edge case I wasn't aware of. Not worth now for sure, because of #4.

Reply:
fair ! :)